lib/os/onoff.c

```diff
@@ -0,0 +1,627 @@
+		rv = -EAGAIN;
+		goto out;
+	}
+
+	if (is_anonymous) {
+		/* Capture the new number of anonymous references and
+		 * verify it won't cause a reference overflow.
+		 */
+		anonymous_count = get_anonymous_count(srv->flags);
+		anonymous_count += 1U;
+		if (anonymous_count > MAX_ANONYMOUS_COUNT) {
+			rv = -EAGAIN;
+			goto out;
+		}
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_TO_OFF:
+		/* Reject if no-wait because operation will not be
+		 * completed when this call returns).
+		 */
+		if (no_wait) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Queue to start after release */
+		__ASSERT_NO_MSG(srv->releaser != NULL);
+		add_client = true;
+		rv = 3;
+		break;
+	case SERVICE_STATE_OFF:
+		/* Reject if in a no-wait context and start could
+		 * wait.
+		 */
+		if ((no_wait || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Start with first request while off */
+		__ASSERT_NO_MSG(srv->refs == 0);
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+		add_client = true;
+		rv = 2;
+		break;
+	case SERVICE_STATE_TO_ON:
+		/* Reject if no-wait because operation will not be
+		 * completed when this call returns.
+		 */
+		if (no_wait) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Already starting, just queue it */
+		add_client = true;
+		rv = 1;
+		break;
+	case SERVICE_STATE_ON:
+		/* Just increment the reference count */
+		notify = true;
+		break;
+	default:
+		rv = -EINVAL;
+		break;
+	}
+
+out:
+	if (add_client) {
+		if (is_anonymous) {
+			srv->flags = set_anonymous_count(srv->flags, anonymous_count);
+		} else {
+			sys_slist_append(&srv->clients, &cli->node);
+		}
+	} else if (notify) {
+		srv->refs += 1;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (start) {
+		__ASSERT_NO_MSG(srv->start != NULL);
+		srv->start(srv, onoff_start_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res)
+{
+	bool notify_clients = false;
+	int client_res = res;
+	bool start = false;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+	struct onoff_client *releaser = srv->releaser;
+
+	/* If the stop operation failed log an error and leave the
+	 * rest of the state in place.
+	 *
+	 * If it succeeded remove the last reference and transition to
+	 * off.
+	 *
+	 * In either case remove the last reference, and notify all
+	 * waiting clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+		notify_clients = true;
+	} else if (sys_slist_is_empty(&clients)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+	} else if (k_is_in_isr()
+		   && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+		notify_clients = true;
```

Comment:
i keep thinking about this case. It can really hit somebody hard. I wonder about optional `k_work` to which start operation would be deferred in that case. what if START_WAITS or STOP_WAITS is set then k_work could be required (although it may be to hard requirement since it can be handled on user side). On the other hand, blocking default work queue for unknown period ... that one is tricky.

Reply:
"This case" being a deferred restart is rejected because it can't be invoked from the ISR?  Everybody who asked for it will get notification of the failure (`-EWOULDBLOCK`), and can request again from thread context.  I don't see any reason to have this service try to provide a solution.

Reply:
The problem is that without looking into implementation user does not know that this can happen. `result` field in `onoff_client` structure is not documented. It should state what kind of errors it may contain apart from service specific.